certificate-manager/src/main/java/io/strimzi/certs/IpAndDnsValidation.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.certs;
+
+import java.util.Locale;
+import java.util.regex.Pattern;
+
+/**
+ * Helper class to help with validation of IP addresses and DNS addresses
+ */
+public class IpAndDnsValidation {
+    // Pattern for matching IPv4 address (requires additional check for segments to be equal or less than 255 - this is done in the isValidIpv4Address method)
+    private static final Pattern IPV4_ADDRESS = Pattern.compile("\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}");
+
+    // Pattern for uncompressed IPv6 address -> case-insensitive, tolerates leading 0s
+    private static final Pattern IPV6_ADDRESS = Pattern.compile("^(?:[\\da-fA-F]{1,4}:){7}[\\da-fA-F]{1,4}$");
+
+    // Pattern for compressed IPv6 address -> case-insensitive, tolerates leading 0s
+    private static final Pattern IPV6_COMPRESSED_ADDRESS = Pattern.compile("^((?:[\\dA-Fa-f]{1,4}(?::[\\dA-Fa-f]{1,4})*)?)::((?:[\\dA-Fa-f]{1,4}(?::[\\dA-Fa-f]{1,4})*)?)$");
+
+    // Pattern for DNS names
+    private static final Pattern DNS_NAME = Pattern.compile("^(" +
+            // a single char dns name
+            "[a-zA-Z\\d]|" +
+            // can't begin or end with -                followed by more labels of same
+            "[a-zA-Z\\d][a-zA-Z\\d\\-]{0,61}[a-zA-Z\\d])(\\.([a-zA-Z\\d]|[a-zA-Z\\d][a-zA-Z\\d\\-]{0,61}[a-zA-Z\\d]))*$");
+
+    // Pattern used to normalize the segments of the IPv6 address
+    private static final Pattern IPV6_LEADING_ZEROS_TRIMMING = Pattern.compile("(^|:)0+([\\da-f]+)(:|$)");
```

Comment:
A segment cannot be all zeros?

Reply:
This is used in the normalization. In general, the segment can be `:0:`, `:00:` or `:0000:` and it all means the same. Similarly, you can be `:00b3`, :`0b3:` or `:b3:` and they are the same. So the normalization removes the leading zeros to end up always with e.g. `:0:` or `:b3`. This allows us to compare the addresses and decide if new cert is needed.

Reply:
But don't _think_ this will match `:00:` because of the `[\\da-f]+`. The `0+` matches greedily, so the `\\d` doesn't match the 2nd `0`.

Reply:
I have test with `00` I can add one more with `000` as well if needed. It seems to work fine. My understanding of the reg-exp is that yes, `0+` matches greedily. But the `[\\da-f]+` does as well. So if all the zeros would go to `0+`, it would not match the expression at all. So when you have only multiple zeroes and nothing else, the last zero always goes to `[\\da-f]+`. E.g. with `:000:`, the first two zeroes will go to `0+` and the last one stays with `[\\da-f]+`.

Reply:
I added the test here: https://github.com/scholzj/strimzi-kafka-operator/blob/6ad18065ba278796984e23442fad5a9aca012c3d/certificate-manager/src/test/java/io/strimzi/certs/IpAndDnsValidationTest.java#L66

Reply:
OK, I believe you. I must have messed up my "test" (which was to use the regex search in IDEA).